src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -6,240 +6,204 @@
 License
 MIT
 """
+from __future__ import annotations
 
-from abc import ABC
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
 from datetime import datetime
-from typing import Tuple
+from typing import Optional, Tuple
 
 import frequenz.api.microgrid.microgrid_pb2 as microgrid_pb
 import pytz
 
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
+@dataclass(frozen=True)
+class ComponentData(ABC):
     """A private base class for strongly typed component data classes."""
 
-    _raw: microgrid_pb.ComponentData
+    component_id: int
+    timestamp: datetime
+    _raw: Optional[microgrid_pb.ComponentData] = field(default=None, init=False)
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    @staticmethod
+    def _set_raw(obj: ComponentData, raw: microgrid_pb.ComponentData) -> None:
+        """Store raw protobuf message.
 
-        Returns:
-            Component id of the message.
-        """
-        return self._raw.id
-
-    @property
-    def timestamp(self) -> datetime:
-        """Return the timestamp of the message.
-
-        Returns:
-            Timestamp of the message.
-        """
-        return self._raw.ts.ToDatetime(tzinfo=pytz.UTC)
-
-
-class MeterData(_BaseComponentData):
-    """A wrapper class for holding meter data."""
-
-    def __init__(self, raw: microgrid_pb.ComponentData) -> None:
-        """Create a `MeterData` instance.
+        Keep the dataclass frozen, but bypass it here to store the raw protobuf
+            message for in case it was needed later.
 
         Args:
+            obj: object to add the raw protobuf message to
             raw: raw component data as decoded from the wire.
         """
-        self._raw = raw
+        obj.__dict__["_raw"] = raw
```

Comment:
Why like this?
Why raw can't be passed in constructor?

Reply:
Because the idea was to exclude it from the constructor. If you want to instantiate from raw protobuf message, you would you `BatteryData.from_proto(raw)` instead.

Reply:
`_raw` has to be optional, because we want to use these structs for the output from `resampling actor`, and it would make tests easier.

We also want to put `_raw` in the base class (`ComponentData`). Because these are all data classes with default `__init__` methods,   If we do that and take it as an optional constructor parameter,  it means all derived classes will have broken default constructors i.e params with default arguments appearing before params without default arguments.

The easy workaround for all of these is to have a special method for setting the `_raw` param.

Reply:
This workaround should be very well documented, because is a hack that on the first sight look terribly wrong, is not trivial to realize why this is needed.

Reply:
Added more verbose documentation.

Reply:
Great, thanks!